[CI] Recover CI job failures

withsmilo opened the pull request:
Clipper CI has always failed since May 17, 2019. To resolve this problem,

* integration tests
  * Fix clipper_admin_tests.py in integration-tests.
  * Fix clipper_metric_docker.py in integration-tests.
* unit tests
  * Fix query_frontend_tests' bug.
* Dockerfile
  * Add six==1.12.* to FrontendExporterDockerfile.
  * Change pyyaml and requests version
    * pyyaml : ==3.12.* -> >=4.2b1
    * requests : ==2.18.* -> ==2.20.0
  * Update mxnet's version on ClipperPy35DevDockerfile. (1.1.* -> 1.4.*)
  * Upgrade pip to latest when building some Docker images.
* build system
  * split all tests with three parts
    * unittest
    * integration_py2
    * integration_py3
* improvements
  * Check Docker container's status after running it.
  * Increase sleep time after starting the metric server in rpc.py. (1s -> 5s)
  * Upgrade Prometheus version from `v2.1.0` to `v2.9.2`.

Commit:
Increase SLO times in clipper_admin_tests

I found that The CI tests using SLO(30ms) might be failed in rare cases.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1974/
Test FAILed.

withsmilo commented:
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1974/ shows this error,
```
[pytorch-container] The command '/bin/sh -c pip install -q torch==1.0.* torchvision==0.2.*' returned a non-zero code: 1 
CI_build.Makefile:472: recipe for target 'pytorch-container' failed
make: *** [pytorch-container] Error 1
```
Temporary network issue?

withsmilo commented:
Jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1976/
Test FAILed.

withsmilo commented:
Jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1978/
Test FAILed.

rkooo567 commented:
Based on the fact that our main build is also failing, there might be some issues in our code? (like dependency issue).

rkooo567 reviewed (dismissed):
Looks good to me!

withsmilo commented:
@rkooo567 I don't know why. Let's retry Jenkins again.

withsmilo commented:
Jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1981/
Test FAILed.

withsmilo commented:
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1981/ shows this error,
```
 [integration_py2_docker_metric] 19-05-18:12:58:11 INFO     [clipper_metric_docker.py:107] Test 1 Passed 
 [integration_py2_docker_metric] 19-05-18:12:58:11 INFO     [clipper_metric_docker.py:109] Test 2: Checking Model Container Metrics 
 [integration_py2_docker_metric] 19-05-18:12:58:11 INFO     [clipper_metric_docker.py:43] Querying: http://localhost:15925/api/v1/series?match[]=clipper_mc_parse_time_ms_sum 
 [integration_py2_docker_metric] 19-05-18:12:58:11 INFO     [clipper_metric_docker.py:46] {u'status': u'success', u'data': []} 
```
Returned data is empty.

Commit:
Add a retry routine to clipper_metric_docker

https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1981/ shows this error,
```
 [integration_py2_docker_metric] 19-05-18:12:58:11 INFO     [clipper_metric_docker.py:107] Test 1 Passed
 [integration_py2_docker_metric] 19-05-18:12:58:11 INFO     [clipper_metric_docker.py:109] Test 2: Checking Model Container Metrics
 [integration_py2_docker_metric] 19-05-18:12:58:11 INFO     [clipper_metric_docker.py:43] Querying: http://localhost:15925/api/v1/series?match[]=clipper_mc_parse_time_ms_sum
 [integration_py2_docker_metric] 19-05-18:12:58:11 INFO     [clipper_metric_docker.py:46] {u'status': u'success', u'data': []}
```
Returned data is empty.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1982/
Test FAILed.

Commit:
Incrase sleep time in clipper_metric_docker

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1983/
Test FAILed.

Commit:
Check container's status after running it

withsmilo commented:
`flatten_json` library was updated to `0.1.7` on May 16, 2019. (https://pypi.org/project/flatten-json/#history) @rkooo567 @simon-mo, I think this update caused the following error and made all Jenkins jobs after May 16 to fail.
```
Traceback (most recent call last):
  File "/usr/src/app/front_end_exporter.py", line 2, in <module>
    from flatten_json import flatten
  File "/usr/local/lib/python3.6/site-packages/flatten_json.py", line 14, in <module>
    import six
ModuleNotFoundError: No module named 'six'
```

Commit:
Add six==1.12.* to FrontendExporterDockerfile

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1984/
Test FAILed.

withsmilo commented:
During `integration_py2_docker_metric` test, Promethus shows this error,
```
level=info ts=2019-05-18T16:16:09.7479649Z caller=main.go:585 msg="Loading configuration file" filename=/etc/prometheus/prometheus.yml
level=error ts=2019-05-18T16:16:09.75432Z caller=main.go:450 msg="Error reloading config" err="couldn't load configuration (--config.file=/etc/prometheus/prometheus.yml): parsing YAML file /etc/prometheus/prometheus.yml: yaml: control characters are not allowed"
level=warn ts=2019-05-18T16:17:03.886925Z caller=main.go:361 msg="Received SIGTERM, exiting gracefully..."
```

Commit:
Change pyyaml and requests version

* pyyaml : ==3.12.* -> >=4.2b1
* requests : ==2.18.* -> ==2.20.0

rkooo567 commented:
I will try to debug our main build later today if ur fix doesn’t alleviate the problem

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1985/
Test FAILed.

withsmilo commented:
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1985/ shows this error,
```
 [integration_py3_mxnet] Exception ignored in: <bound method MXDataIter.__del__ of <mxnet.io.MXDataIter object at 0x7fe6218b8748>> 
 [integration_py3_mxnet] Traceback (most recent call last): 
 [integration_py3_mxnet] File "/usr/local/lib/python3.5/dist-packages/mxnet/io.py", line 803, in __del__ 
 [integration_py3_mxnet] AttributeError: 'NoneType' object has no attribute 'MXDataIterFree'
```

Commit:
Update mxnet's version on ClipperPy35DevDockerfile

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1986/
Test FAILed.

withsmilo commented:
Jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1987/
Test FAILed.

withsmilo commented:
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1986/ shows this error,
```
 [pytorch-container] Could not install packages due to an EnvironmentError: HTTPSConnectionPool(host='files.pythonhosted.org', port=443): Max retries exceeded with url: /packages/93/b3/672813e65ac7605bac14a2a8e40e1a18e03bf0ac588c6a799508ee66c289/torch-1.0.1.post2-cp27-cp27mu-manylinux1_x86_64.whl (Caused by ReadTimeoutError("HTTPSConnectionPool(host='files.pythonhosted.org', port=443): Read timed out. (read timeout=15)",)) 
 [pytorch-container]  
 [pytorch-container] You are using pip version 10.0.1, however version 19.1.1 is available. 
 [pytorch-container] You should consider upgrading via the 'pip install --upgrade pip' command. 
```

Commit:
Upgrade pip when installing

```
 [pytorch-container] Could not install packages due to an EnvironmentError: HTTPSConnectionPool(host='files.pythonhosted.org', port=443): Max retries exceeded with url: /packages/93/b3/672813e65ac7605bac14a2a8e40e1a18e03bf0ac588c6a799508ee66c289/torch-1.0.1.post2-cp27-cp27mu-manylinux1_x86_64.whl (Caused by ReadTimeoutError("HTTPSConnectionPool(host='files.pythonhosted.org', port=443): Read timed out. (read timeout=15)",))
 [pytorch-container]
 [pytorch-container] You are using pip version 10.0.1, however version 19.1.1 is available.
 [pytorch-container] You should consider upgrading via the 'pip install --upgrade pip' command.
```

Commit:
Select randomized port for prometheus

```
 [integration_py2_many_apps_many_models] 19-05-19:20:27:48 INFO     [test_utils.py:147] Problem starting Clipper: 500 Server Error: Internal Server Error ("driver failed programming external connectivity on endpoint metric_frontend-15123 (971f668722afe5dfe1e6a773a0bc38ec3a2c1a8dce3a5772eedf5a9b9d99eed8): Bind for 0.0.0.0:9090 failed: port is already allocated")
```

Commit:
When a integration test is failed, sleep randomized time

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1988/
Test FAILed.

Commit:
Revert "Add a retry routine to clipper_metric_docker"

This reverts commit 665168087c02844948d3de3659dec0aebc34f44d.

Commit:
increase sleep time in rpc.py

Commit:
Increase sleep time in clipper_metric_docker.py

Commit:
Upgrade Prom's version

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1989/
Test PASSed.

withsmilo commented:
Finally, Jenkins job is passed. I will retry it three more times to check stability.

withsmilo commented:
Jenkins test this please

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1990/
Test PASSed.

withsmilo commented:
Jenkins test this please

rkooo567 commented on clipper_admin/clipper_admin/docker/docker_container_manager.py:
```diff
@@ -475,6 +476,19 @@ def stop_all(self, graceful=True):
+           (inspected is not None and inspected.get("State").get("Health").get("Status") == "healthy"):
+            return
+        else:
+            msg = "{} container is not running yet".format(name)
```
Since we have lots of retry (max 5 minutes for each container status check), I think it is better having a little more descriptive message.

Why don't we change the msg to 
`{} container is not running yet or broken. We will try to run again. Please analyze logs if it keeps failing.`?

withsmilo replied:
Good advice. I will update the message.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1991/
Test PASSed.

withsmilo commented:
Jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1992/
Test FAILed.

rkooo567 commented:
The latest fail is due to 
```
[unittest_frontend] [03:32:02.496][info] [STATEDB] Persistent state DB created

[unittest_frontend] terminate called after throwing an instance of 'zmq::error_t'

[unittest_frontend] what(): Address already in use

[unittest_frontend] /clipper/bin/run_unittests.sh: line 117: 11 Aborted (core dumped) ./src/frontends/frontendtests --redis_port $REDIS_PORT
```

I will run a test again

rkooo567 commented:
Jenkins test this please

Commit:
Change error log in docker_container_manager

withsmilo commented:
@rkooo567, After https://amplab.cs.berkeley.edu/jenkins/job/Clipper-PRB/1993/ finished, I will push new commits, which contain your review and fix the `core dump` problem.

Commit:
Add debug log to run_unittests.sh

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1993/
Test FAILed.

Commit:
Fix query_frontend_tests' bug

rkooo567 commented:
New changes look good to me!

rkooo567 commented:
@simon-mo @withsmilo  I will debug keras deployer / RBAC PR once this is merged

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1994/
Test FAILed.

withsmilo commented:
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1994/ shows, all the integration tests(
integration_py2_pyspark_pipeline, integration_py2_tensorflow, integration_py2_pyspark, integration_py2_many_apps_many_models, integration_py3_pyspark, integration_py3_pysparkml, integration_py2_admin_unit_test) were destroyed simultaneously. I don't know why. (maybe limited cluster resources?) Let's retry it.

withsmilo commented:
Jenkins test this please

withsmilo commented:
@rkooo567 @simon-mo 
How about split all the tests with three parts?
```
make -j10 -f CI_test.Makefile unittest
make -j10 -f CI_test.Makefile integration_py2
make -j10 -f CI_test.Makefile integration_py3
```

Note that current setting is like,
```
make -j10 -f CI_test.Makefile all
```

rkooo567 commented:
I think many of our CI issues are due to parallel unit tests, so I think it is worth trying until @simon-mo fixes CI one more time?

Commit:
split all tests with three parts

* unittest
* integration_py2
* integration_py3

simon-mo commented:
sounds good to me! thank you for your time and effort.

withsmilo commented:
@rkooo567 @simon-mo Thanks for your quick response. After https://amplab.cs.berkeley.edu/jenkins/job/Clipper-PRB/1995/ finished, I will push new commit.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1995/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1996/
Test PASSed.

withsmilo commented:
Jenkins test this please

withsmilo commented:
I will run it three more times to check stability.

rkooo567 reviewed (approved):
Looks good to me!!

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1997/
Test PASSed.

withsmilo commented:
Jenkins test this please

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1998/
Test PASSed.

withsmilo commented:
Jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1999/
Test FAILed.

withsmilo commented:
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1999/ shows,
```
===== start: integration_py2_mxnet =====
(    python ./bin/retry_with_timeout.py --retry 1 --timeout 40m --     docker run --rm --network=host -v /var/run/docker.sock:/var/run/docker.sock -v /tmp:/tmp         -e CLIPPER_REGISTRY=clippertesting         -e CLIPPER_TESTING_DOCKERHUB_PASSWORD=LIPPER_TESTING_DOCKERHUB_PASSWORD         clippertesting/unittests:3d00a375d3         "python /clipper/integration-tests/deploy_mxnet_models.py" ) 2>&1 | python3 ./bin/colorize_output.py --tag integration_py2_mxnet
...
 [integration_py2_mxnet] Starting Trial 0 with timeout 2400.0 seconds 
 [integration_py2_mxnet] Sleep 414 
 [integration_py2_mxnet] Starting Trial 1 with timeout 2400.0 seconds 
 [integration_py2_mxnet] Sleep 416 
 [integration_py2_mxnet] All retry failed. 
CI_test.Makefile:152: recipe for target 'integration_py2_mxnet' failed
make: *** [integration_py2_mxnet] Error 1
make: *** Waiting for unfinished jobs....
```
integration_py2_mxnet didn't start at all.

Commit:
Sleep random time before a test

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/2000/
Test PASSed.
